kedro/framework/context/context.py

```diff
@@ -159,65 +160,45 @@ def _update_nested_dict(old_dict: Dict[Any, Any], new_dict: Dict[Any, Any]) -> N
                 old_dict[key] = value
 
 
+def _expand_full_path(project_path: Union[str, Path]) -> Path:
+    return Path(project_path).expanduser().resolve()
+
+
+@define
```

Comment:
Just from reading the docs, I wonder if we should be using https://www.attrs.org/en/stable/api.html#attr.attrs.frozen
```suggestion
@attrs.frozen
```
Maybe cleaner overall since it allows you to remove the `on_setattr=attrs.setters.frozen`? Or would it interfere with `_package_name` etc. attributes and then you have to use `field` for them?

Reply:
2 reasons for this.

1. To keep the API identical as before, so it is a pure refactor work that doesn't change any interface.
2. Initially I find `@frozen` doesn't work with the transforming logic with `__attrs_post_init__`, but I now find that it can works with `converter` so we could make this entire class immutable.

Reply:
But let's go with `@frozen` now, it is more cleaner and actually simplifies the test as well as we got 1 `FrozenInstsance` error to check only.